### PR TITLE
Add specific exceptions for invalid entity types and location levels

### PR DESCRIPTION
**Pull request by alexispeter**
cm-changes for https://github.com/cargomedia/sk/issues/4355

### CODE REVIEW

File: library/CM/Model/Entity/Abstract.php [hunk 1]
```diff
@@ -50,9 +50,14 @@ public function getUserId() {
      * @param int $type
      * @param int $id
      * @return CM_Model_Entity_Abstract
+     * @throws CM_Entity_InvalidTypeException
      */
     final public static function factory($type, $id) {
         $className = self::_getClassName($type);
+        $calledClass = get_called_class();
+        if (!is_subclass_of($className, $calledClass)) {
```
Comment (njam) [inline library/CM/Model/Entity/Abstract.php]: Could we do `$className instanceof static`?

Comment (njam) [inline library/CM/Model/Entity/Abstract.php]: …and to print `static::class`?

Comment (alexispeter) [inline library/CM/Model/Entity/Abstract.php]: `instanceof` only works with objects as far as I know.

Comment (alexispeter) [inline library/CM/Model/Entity/Abstract.php]: You could replace `get_called_class()` with `static::class` if there was any reason for doing so.

**Pull request merged**